kbs/docs/id-key.md

```diff
@@ -0,0 +1,84 @@
+# ID-key Plugin
+
+In addition to fetchable resources upon successful attestation, KBS also offers the ID-key plugin for encrypting and decrypting data with a key provisioned by the KBS. This key can be persisted across multiple reboots of CVMs and help tie decryption resources to a specific KBS instance.
+
+## Data Encryption
+
+As a guest owner would be responsible for encrypting persistent storage, they are the sole authority that is permitted to request resources be encrypted. The KBS is expected to be fully in control by the guest owner with credentials configured.
+
+## Flow
+
+### Provisioning encrypted data
+
+A credentialed guest owner is the only one authorized to request that keys be encrypted by the ID-key plugin. To do this, the guest owner will POST some plaintext data (created locally) to be encrypted. This plaintext data could be a LUKS passphrase, SecureBoot keys, etc. The actual purpose of the data is of no concern to the KBS.
+
+`POST /kbs/v0/id-key/{base64-plaintext-data}`
+
+
+```
+          ┌───────────┐        
+          │   POST    │        
+Owner ──► ├───────────┤ ──► KBS
+  ▲       │ Plaintext │        
+  │       │ data      │      │ 
+  │       └───────────┘      │ 
+  │                          │ 
+  │                          │ 
+  │       ┌───────────┐      │ 
+  └────── │ Encrypted │ ◄────┘ 
+          │ data      │        
+          └───────────┘        
+```
+
+With this, the guest owner is now free to configure persistent storage as they wish for their CVMs to have access to encrypted data when they boot. Upon successful attestation, this encrypted data will be presented to the KBS by the CVM for decryption.
+
+### Post-attestation, decrypting data
+
+With a CVM booted and attested, it is now assumed that the CVM was configured in such a way that it has access to readable storage that contains the encrypted data. The CVM can now send this data to the KBS for decryption.
+
+To prevent replay attacks, the CVM must wrap the encrypted data in a ECDH key cipher. To support this, the ID-key plugin presents an endpoint `ecdh-pub-sec1` that a VM can use to fetch the EC public key in order to wrap the encrypted data.
+
+`GET /kbs/v0/id-key/ecdh-pub-sec1`
```

Comment:
`ecdh-pub-sec1` seems to be a magic string on the `/kbs/v0/id-key/{string}` route, should we move it to another hierarchy?

Reply:
To prevent replay attacks, the encrypted data is wrapped with an ECDH cipher. To build this cipher, the client needs the server's EC public key. This endpoint allows for fetching that public key.

I chose SEC1 encoding because its a "standard" format for encoding EC public keys, found in _Elliptic-Curve-Point-to-Octet-String encoding described in [SEC 1: Elliptic Curve Cryptography (Version 2.0) section 2.3.3 (page 10)](http://www.secg.org/sec1-v2.pdf)_.

What exactly do you mean by "hierarchy"? I think the existing way it's done is fine, but am open to suggestions if you think another would be an improvement.

Reply:
ah, this is also just a question about API design. it's not optimal to have a string with special meaning on the same hiearchy as resource descriptions: like `/apple/count-apples` and `/apple/1`, `/apple/2`, etc....

Reply:
Oh ok, I can modify for the public key to be fetched at:
`/kbs/v0/id-key/ecdh-pub-sec1`

While the key decryption is at an endpoint like:
`/kbs/v0/id-key/decrypt/...`

Would this be preferred?